### ISSUE TITLE
Compat API endpoint for "Inspect Container" response "Image" field should contain ImageID.

### ISSUE THREAD

**Issue by ashleysommer** [containers/podman#15832]
/kind bug

**Description**

The compat API endpoint to inspect container `/v1.40/containers/{container_name}/json` response has a filed named "Image", on Docker it contains the image hash, but on Podman, it contains the image name. This causes one or more operations in Docker-PY module to fail.

_Note_, the compat container _list_ endpoint `/v1.40/containers/json` _does_ include the image ID in field `ImageID`, so its omission in the single container inspect endpoint is a bug.

**Steps to reproduce the issue:**

1. `podman pull alpine:3.16`

2. `podman run --name my_container -i -t alpine:3.16 /bin/true`  (don't include `--rm`)

4. `podman container inspect my_container` 

5. `curl --unix-socket /run/user/1000/podman/podman.sock -H accept:application/json http://localhost/v1.40/containers/my_container/json`

**Describe the results you received:**
Image field in the response shows image name:
```
{
   "Id":"02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7",
   "Created":"2022-09-16T08:51:15.331628601Z",
   "Path":"/bin/true",
   "Args":[
      "/bin/true"
   ],
   "State":{
      "Status":"exited",
      "Running":false,
      "Paused":false,
      "Restarting":false,
      "OOMKilled":false,
      "Dead":false,
      "Pid":0,
      "ExitCode":0,
      "Error":"",
      "StartedAt":"2022-09-16T08:51:15.421336931Z",
      "FinishedAt":"2022-09-16T08:51:15.419473831Z",
      "Health":{
         "Status":"",
         "FailingStreak":0,
         "Log":null
      }
   },
   "Image":"docker.io/library/alpine:3.16",
   "ResolvConfPath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/resolv.conf",
   "HostnamePath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/hostname",
   "HostsPath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/hosts",
   "LogPath":"",
   "Name":"/my_container",
   "RestartCount":0,
   "Driver":"overlay",
   "Platform":"linux",
   "MountLabel":"",
   "ProcessLabel":"",
   "AppArmorProfile":"",
   "ExecIDs":[
      
   ],
   "HostConfig":{
     ...
   }
}
```

**Describe the results you expected:**
Image field in the response should be "sha256:xyzhashxyz"
```
{
   "Id":"02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7",
   "Created":"2022-09-16T08:51:15.331628601Z",
   "Path":"/bin/true",
   "Args":[
      "/bin/true"
   ],
   "State":{
      "Status":"exited",
      "Running":false,
      "Paused":false,
      "Restarting":false,
      "OOMKilled":false,
      "Dead":false,
      "Pid":0,
      "ExitCode":0,
      "Error":"",
      "StartedAt":"2022-09-16T08:51:15.421336931Z",
      "FinishedAt":"2022-09-16T08:51:15.419473831Z",
      "Health":{
         "Status":"",
         "FailingStreak":0,
         "Log":null
      }
   },
   "Image":"9c6f0724472873bb50a2ae67a9e7adcb57673a183cea8b06eb778dca859181b5",
   "ResolvConfPath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/resolv.conf",
   "HostnamePath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/hostname",
   "HostsPath":"/run/user/1000/containers/overlay-containers/02e2d9d92829389c0a3d012effb932cd09687f9a2995c52c5ff3a7d3817d9ae7/userdata/hosts",
   "LogPath":"",
   "Name":"/my_container",
   "RestartCount":0,
   "Driver":"overlay",
   "Platform":"linux",
   "MountLabel":"",
   "ProcessLabel":"",
   "AppArmorProfile":"",
   "ExecIDs":[
      
   ],
   "HostConfig":{
       ...
   }
}
```

_Note_, the compat container _list_ endpoint `/v1.40/containers/json` _does_ include the image ID in filed `ImageID`, so it is not affected by this bug.

This is the section of code in Docker-Py library that crashes:
https://github.com/docker/docker-py/blob/923e067dddc3d4b86e4e620a99fcdcdafbd17a98/docker/models/containers.py#L37-L40
Notice it is expecting an ImageID in the form `sha256:xyzhashxyz`, but it does not find it, so it falls back to field `Image` then splits the value `alpine:3.16` on the colon, returning just "3.16" instead of the image hash, this  causes the image lookup to fail.

Minimal reproduction in Python:
```python
import docker
client = docker.from_env()
container = client.containers.get("my_container")
image = container.image    # <-- crashes here
```
error:
```python
docker.errors.ImageNotFound: 404 Client Error for http+docker://localhost/v1.40/images/3.16/json: Not Found ("failed to find image 3.16: docker.io/library/3.16: No such image")
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
podman version 4.1.1
```

**Output of `podman info`:**

```
host:
  arch: amd64
  buildahVersion: 1.26.1
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.1.3-1
    path: /usr/bin/conmon
    version: 'conmon version 2.1.3, commit: ab52a597278b20173440140cd810dc9fa8785c93'
  cpuUtilization:
    idlePercent: 97.96
    systemPercent: 0.37
    userPercent: 1.67
  cpus: 16
  distribution:
    distribution: manjaro
    version: unknown
  eventLogger: journald
  hostname: manjryzen
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.18.11-xanmod1-MANJARO
  linkmode: dynamic
  logDriver: journald
  memFree: 511868928
  memTotal: 16701726720
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: /usr/bin/crun is owned by crun 1.5-1
    path: /usr/bin/crun
    version: |-
      crun version 1.5
      commit: 54ebb8ca8bf7e6ddae2eb919f5b82d1d96863dea
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /etc/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.2.0-1
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.4
  swapFree: 11046121472
  swapTotal: 17179865088
  uptime: 105h 17m 13.89s (Approximately 4.38 days)
plugins:
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/flubba86/.config/containers/storage.conf
  containerStore:
    number: 25
    paused: 0
    running: 2
    stopped: 23
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/flubba86/.local/share/containers/storage
  graphRootAllocated: 772024565760
  graphRootUsed: 249766158336
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 131
  runRoot: /run/user/1000/containers
  volumePath: /home/flubba86/.local/share/containers/storage/volumes
version:
  APIVersion: 4.1.1
  Built: 1659559968
  BuiltTime: Thu Aug  4 06:52:48 2022
  GitCommit: f73d8f8875c2be7cd2049094c29aff90b1150241-dirty
  GoVersion: go1.19
  Os: linux
  OsArch: linux/amd64
  Version: 4.1.1
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
Manjaro official podman package, v4.1.1
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

No changes in behaviour between my version and latest codebase.

**Comment by vrothberg** [comment]
Thank you for reaching out, @ashleysommer!

I can reproduce using the current HEAD.  @boaz0, interested in tackling the bug?

**Comment by ashleysommer** [comment]
@vrothberg 
I've just realized I made a mistake. The API response type for Container Inspect for the Docker API v1.40 does not include an "ImageID" field.
See the definition:
https://github.com/moby/moby/blob/master/api/types/types.go#L449-L481


After going through the relevant Podman code and Docker code, I see where the misunderstanding lies.

This is where Docker populates the `Image` field, it uses the image hash:
https://github.com/moby/moby/blob/a263241ea16f1b018fc1c8f5c6564d84f1153543/daemon/inspect.go#L170

Podman does the same, correctly:
```
https://github.com/containers/podman/blob/4ba6ee306088a4d322fe28f292bf5d5c4d17b778/pkg/api/handlers/compat/containers.go#L491
```
So I'm not sure where this bug is being manifested. Somewhere the "Image" field is being overwritten with  `imageName`. I'm not familiar enough Go lang to investigate any further.

**Comment by mheon** [comment]
I think this might be fixed already in main.

**Comment by ashleysommer** [comment]
Hmm, you're right.
https://github.com/containers/podman/commit/6411f57e789ecef322c120d34ad634495e54d257

I checked for recent commits in that file because I thought it seems like something that might've been changed (behaviour in code is different to behaviour I saw in the release), but when I checked yesterday I didn't see anything recent around that area, and I even checked git blame on that line of code, but didn't see anything. But today I checked again and its pretty obvious there was a commit last month in master that fixed it.

Fixed by https://github.com/containers/podman/commit/6411f57e789ecef322c120d34ad634495e54d257
